storm-server/src/main/java/org/apache/storm/metricstore/rocksdb/RocksDbStore.java

```diff
@@ -0,0 +1,636 @@
+    private BlockingQueue queue = new LinkedBlockingQueue(MAX_QUEUE_CAPACITY);
+    private RocksDbMetricsWriter metricsWriter = null;
+    private MetricsCleaner metricsCleaner = null;
+    private Meter failureMeter = null;
+
+    interface RocksDbScanCallback {
+        boolean cb(RocksDbKey key, RocksDbValue val);  // return false to stop scan
+    }
+
+    /**
+     * Create metric store instance using the configurations provided via the config map.
+     *
+     * @param config Storm config map
+     * @throws MetricException on preparation error
+     */
+    public void prepare(Map config) throws MetricException {
+        validateConfig(config);
+
+        this.failureMeter = StormMetricsRegistry.registerMeter("RocksDB:metric-failures");
+
+        RocksDB.loadLibrary();
+        boolean createIfMissing = ObjectReader.getBoolean(config.get(DaemonConfig.STORM_ROCKSDB_CREATE_IF_MISSING), false);
+
+        try (Options options = new Options().setCreateIfMissing(createIfMissing)) {
+            // use the hash index for prefix searches
+            BlockBasedTableConfig tfc = new BlockBasedTableConfig();
+            tfc.setIndexType(IndexType.kHashSearch);
+            options.setTableFormatConfig(tfc);
+            options.useCappedPrefixExtractor(RocksDbKey.KEY_SIZE);
+
+            String path = getRocksDbAbsoluteDir(config);
+            LOG.info("Opening RocksDB from " + path);
+            db = RocksDB.open(options, path);
+        } catch (RocksDBException e) {
+            String message = "Error opening RockDB database";
+            LOG.error(message, e);
+            throw new MetricException(message, e);
+        }
+
+        // create thread to delete old metrics and metadata
+        Integer retentionHours = Integer.parseInt(config.get(DaemonConfig.STORM_ROCKSDB_METRIC_RETENTION_HOURS).toString());
+        Integer retentionPeriod = 0;
+        if (config.containsKey(DaemonConfig.STORM_ROCKSDB_METRIC_RETENTION_PERIOD_HOURS)) {
+            retentionPeriod = Integer.parseInt(config.get(DaemonConfig.STORM_ROCKSDB_METRIC_RETENTION_PERIOD_HOURS).toString());
+        }
+        metricsCleaner = new MetricsCleaner(this, retentionHours, retentionPeriod, failureMeter);
+
+        // create thread to process insertion of all metrics
+        metricsWriter = new RocksDbMetricsWriter(this, this.queue, this.failureMeter);
+
+        int cacheCapacity = Integer.parseInt(config.get(DaemonConfig.STORM_ROCKSDB_METADATA_STRING_CACHE_CAPACITY).toString());
+        StringMetadataCache.init(metricsWriter, cacheCapacity);
+        readOnlyStringMetadataCache = StringMetadataCache.getReadOnlyStringMetadataCache();
+        metricsWriter.init(); // init the writer once the cache is setup
+
+        // start threads after metadata cache created
+        Thread thread = new Thread(metricsCleaner, "RocksDbMetricsCleaner");
+        thread.start();
+        thread = new Thread(metricsWriter, "RocksDbMetricsWriter");
+        thread.start();
```

Comment:
Should we mark these threads as daemons?  If not we should make sure that they shut down when we close the Store.